Add `--verify`

Urhengulas opened the pull request:
This PR adds a `--verify` option to the `BinaryDownloadOptions`. So it gets added to the `download`, `flash` and `run` subcommands.

Testing it on my `nrf52840-dk` the verification shows up in the logs:
```console
$ probe-rs run --chip nRF52840_xxAA --verify target/thumbv7em-none-eabihf/debug/hello
<...>
DEBUG probe_rs::flashing::loader: committing RAM!
DEBUG probe_rs::flashing::loader:     region: 20000000-20040000 (262144 bytes)
DEBUG probe_rs::flashing::loader:      -- empty.
DEBUG probe_rs::flashing::loader: Verifying!
DEBUG probe_rs::flashing::loader:     data: 00000000-00001498 (5272 bytes)
<...>
```

---

Style question: For the other fields of the `BinaryDownloadOptions` you explicitly specify the `name`, `long` and `help`. Is there a reason for that? Usually `clap` just infers this info. So e.g. `restore_unwritten` could be simplified like following without changing the output:

```diff
-#[clap(
-    name = "restore-unwritten",
-    long = "restore-unwritten",
-    help = "Enable this flag to restore all bytes erased in the sector erase but not overwritten by any page."
-)]
+/// Enable this flag to restore all bytes erased in the sector erase but not overwritten by any page.
+#[clap(long)]
pub restore_unwritten: bool,
```

If you prefer the explicit style, I can adapt my PR. If you like the simple style, I can add a commit to adapt the other fields.

Urhengulas commented:
What should I do about the `cargo-deny` failure?

It seems to be due to `rustls-webpki` and `webpki` having a vulnerability. `cargo update` resolves the errors (the warnings persist).

```
error[vulnerability]: rustls-webpki: CPU denial of service in certificate path building
    ┌─ /home/urhengulas/Documents/github.com/probe-rs/probe-rs/Cargo.lock:232:1
    │
232 │ rustls-webpki 0.101.1 registry+https://github.com/rust-lang/crates.io-index
    │ --------------------------------------------------------------------------- security vulnerability detected
    │
    = ID: RUSTSEC-2023-0053
    = Advisory: https://rustsec.org/advisories/RUSTSEC-2023-0053
    = When this crate is given a pathological certificate chain to validate, it will
      spend CPU time exponential with the number of candidate certificates at each
      step of path building.
      
      Both TLS clients and TLS servers that accept client certificate are affected.
      
      We now give each path building operation a budget of 100 signature verifications.
      
      The original `webpki` crate is also affected.
      
      This was previously reported in the original crate
      <https://github.com/briansmith/webpki/issues/69> and re-reported to us
      recently by Luke Malinowski.
    = Solution: Upgrade to >=0.100.2, <0.101.0 OR >=0.101.4 (try `cargo update -p rustls-webpki`)
    = rustls-webpki v0.101.1
      └── rustls v0.21.5
          ├── hyper-rustls v0.24.1
          │   └── reqwest v0.11.18
          │       ├── cmsis-pack v0.6.2
          │       │   └── target-gen v0.20.0
          │       └── target-gen v0.20.0 (*)
          ├── reqwest v0.11.18 (*)
          └── tokio-rustls v0.24.1
              ├── hyper-rustls v0.24.1 (*)
              └── reqwest v0.11.18 (*)

error[vulnerability]: webpki: CPU denial of service in certificate path building
    ┌─ /home/urhengulas/Documents/github.com/probe-rs/probe-rs/Cargo.lock:337:1
    │
337 │ webpki 0.22.0 registry+https://github.com/rust-lang/crates.io-index
    │ ------------------------------------------------------------------- security vulnerability detected
    │
    = ID: RUSTSEC-2023-0052
    = Advisory: https://rustsec.org/advisories/RUSTSEC-2023-0052
    = When this crate is given a pathological certificate chain to validate, it will
      spend CPU time exponential with the number of candidate certificates at each
      step of path building.
      
      Both TLS clients and TLS servers that accept client certificate are affected.
      
      This was previously reported in
      <https://github.com/briansmith/webpki/issues/69> and re-reported recently
      by Luke Malinowski.
      
      `rustls-webpki` is a fork of this crate which contains a fix for this issue
      and is actively maintained.
    = Solution: No safe upgrade is available!
    = webpki v0.22.0
      └── webpki-roots v0.22.6
          └── reqwest v0.11.18
              ├── cmsis-pack v0.6.2
              │   └── target-gen v0.20.0
              └── target-gen v0.20.0 (*)
```

Yatekii commented:
> Is there a reason for that?
No. I think it might have been required at one point when everything was based on structopt or some simple copy pasta abuse :)
Implicit is definitiely better since it avoids typos and the likes.

Yatekii commented:
As for the `webpki` thingy I am really not sure. I would just whitelist it for now.

Yatekii commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added --no-location option to the CLI run command, which suppresses the filename and line number
   information from the rtt log (#1704)
 - target-gen: Add new `--test-address` option to the `target-gen test` subcommand. (#1708)
+- `cli`: Add `--verify` flag to `download`, `flash` and `run`
```
We could add the PR number here :)

Urhengulas commented:
I've added the PR number and added the two advisory IDs to the ignore list. Hopefully CI succeeds now 🙏🏾 

I will refactor the clap arguments in a small follow-up PR 🧹

Yatekii reviewed (approved):
Thanks a lot!
